scripts/gb-model/compose_network.py

```diff
@@ -0,0 +1,311 @@
+# SPDX-FileCopyrightText:  gb-open-market-model contributors
+#
+# SPDX-License-Identifier: MIT
+
+"""
+Compose the Great Britain focused PyPSA network described in
+``doc/gb-model/index.rst``.
+
+The rule assembles the clustered PyPSA-Eur base network with GB-specific
+artefacts (manual region shapes, neighbouring countries, adjusted grid
+connection costs) so that downstream rules can import a consistent
+``networks/composed_{clusters}.nc`` snapshot.
+"""
+
+import copy
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Mapping
+
+import pandas as pd
+import pypsa
+
+from scripts._helpers import configure_logging, set_scenario_config
+from scripts.add_electricity import (
+    attach_hydro,
+    attach_wind_and_solar,
+    load_and_aggregate_powerplants,
+    load_costs,
+    sanitize_carriers,
+    sanitize_locations,
+)
+from scripts.prepare_sector_network import add_electricity_grid_connection
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class CompositionContext:
+    resources_root: Path
+    countries: tuple[str, ...]
+    costs_path: Path | None
+    costs_config: dict[str, Any]
+    max_hours: dict[str, Any] | None
+
+
+def create_context(network_path: str, config: dict[str, Any]) -> CompositionContext:
+    resources_root = Path(network_path).parents[1]
+    costs_cfg = copy.deepcopy(config.get("costs", {}))
+    cost_year = costs_cfg.get("year")
+    costs_path = (
+        resources_root / f"costs_{cost_year}.csv" if cost_year is not None else None
+    )
+
+    return CompositionContext(
+        resources_root=resources_root,
+        countries=tuple(config.get("countries", [])),
+        costs_path=costs_path,
+        costs_config=costs_cfg,
+        max_hours=config.get("electricity", {}).get("max_hours"),
+    )
+
+
+def _load_powerplants(
+    powerplants_path: Path | None,
+    costs,
+    config: dict[str, Any],
+) -> pd.DataFrame:
+    clustering_cfg = config.get("clustering", {})
+    consider_efficiency = clustering_cfg.get("consider_efficiency_classes", False)
+    aggregation_strategies = clustering_cfg.get("aggregation_strategies", {})
+    exclude_carriers = clustering_cfg.get("exclude_carriers", [])
+
+    columns = ["bus", "carrier", "p_nom", "max_hours"]
+
+    if powerplants_path is None:
+        logger.info(
+            "No power plant input provided; continuing without existing capacities"
+        )
+        return pd.DataFrame(columns=columns)
+
+    if not powerplants_path.exists():
+        logger.warning(
+            "Power plant data %s not found; continuing without existing capacities",
+            powerplants_path,
+        )
+        return pd.DataFrame(columns=columns)
```

Comment:
These seem unnecessary. The rule won't trigger without the path being defined and existing...